### PR TITLE
Motion Model / Adapter versatility

### PR DESCRIPTION
**Pull request by Arlaz**
## In the MotionModel and MotionAdapter for AnimateDiff pipeline :
- allow to use a different number of layers per block
- allow to use a different number of transformer per layers per block
- allow a different number of motion attention head per block
- allow to pass the dropout argument

These modifications are needed for some custom trained model that follow the SDXL architecture more closely (less down block) and more Transformer block per layers.

New arguments / tuple instead of int can now be used, as in the following example :

```python
adapter = MotionAdapter(
    block_out_channels=(320, 640, 1280),  # to be checked with the base model unet for compatibility
    motion_layers_per_block=2,  # to be checked with the base model unet for compatibility
    motion_transformer_per_layers=(1, 2, 6),  # free to choose
    motion_mid_block_layers_per_block=1,
    motion_transformer_per_mid_layers=10,
    motion_num_attention_heads=(5, 8, 10), # free to choose
    motion_norm_num_groups=32,
    motion_max_seq_length=32,
    use_motion_mid_block=True,
    conv_in_channels=None,
)
```

Important : this PR does not break any existing code

## Before submitting
- [X] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [X] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sayakpaul @yiyixuxu @DN6

### CODE REVIEW

File: src/diffusers/models/unets/unet_3d_blocks.py [hunk 1]
```diff
@@ -1605,20 +1684,37 @@ def __init__(
         num_attention_heads: int = 1,
         output_scale_factor: float = 1.0,
         cross_attention_dim: int = 1280,
-        dual_cross_attention: float = False,
-        use_linear_projection: float = False,
-        upcast_attention: float = False,
+        dual_cross_attention: bool = False,
```
Comment (DN6) [inline src/diffusers/models/unets/unet_3d_blocks.py:1687]: Thanks for catching this 👍🏽

File: src/diffusers/models/unets/unet_motion_model.py [hunk 2]
```diff
@@ -88,9 +98,11 @@ class MotionAdapter(ModelMixin, ConfigMixin):
     def __init__(
         self,
         block_out_channels: Tuple[int, ...] = (320, 640, 1280, 1280),
-        motion_layers_per_block: int = 2,
+        motion_layers_per_block: Union[int, Tuple[int]] = 2,
+        motion_transformer_per_layers: Union[int, Tuple[int], Tuple[Tuple[int]]] = 1,
```
Comment (DN6) [inline src/diffusers/models/unets/unet_motion_model.py:102]: Small nit. Could we rename to `motion_transformer_layers_per_block`?

File: src/diffusers/models/unets/unet_motion_model.py [hunk 3]
```diff
@@ -88,9 +98,11 @@ class MotionAdapter(ModelMixin, ConfigMixin):
...
         motion_mid_block_layers_per_block: int = 1,
-        motion_num_attention_heads: int = 8,
+        motion_transformer_per_mid_layers: Union[int, Tuple[int]] = 1,
```
Comment (DN6) [inline src/diffusers/models/unets/unet_motion_model.py:104]: Could we rename to `motion_transformer_layers_per_mid_block`

File: src/diffusers/models/unets/unet_motion_model.py [hunk 4]
```diff
@@ -208,20 +258,26 @@ def __init__(
...
         norm_num_groups: int = 32,
         norm_eps: float = 1e-5,
         cross_attention_dim: int = 1280,
         transformer_layers_per_block: Union[int, Tuple[int], Tuple[Tuple]] = 1,
-        reverse_transformer_layers_per_block: Optional[Tuple[Tuple[int]]] = None,
+        reverse_transformer_layers_per_block: Optional[Union[int, Tuple[int], Tuple[Tuple]]] = None,
+        temporal_transformer_layers_per_block: Union[int, Tuple[int], Tuple[Tuple]] = 1,
+        reverse_temporal_transformer_layers_per_block: Optional[Union[int, Tuple[int], Tuple[Tuple]]] = None,
+        transformer_layers_per_midblock: Optional[Union[int, Tuple[int]]] = None,
+        temporal_transformer_layers_per_midblock: Optional[Union[int, Tuple[int]]] = 1,
```
Comment (DN6) [inline src/diffusers/models/unets/unet_motion_model.py:273]:
```suggestion
        transformer_layers_per_mid_block: Optional[Union[int, Tuple[int]]] = None,
        temporal_transformer_layers_per_mid_block: Optional[Union[int, Tuple[int]]] = 1,
```

File: src/diffusers/models/unets/unet_motion_model.py [hunk 5]
```diff
@@ -304,6 +370,20 @@ def __init__(
         if isinstance(transformer_layers_per_block, int):
             transformer_layers_per_block = [transformer_layers_per_block] * len(down_block_types)
 
+        if isinstance(reverse_transformer_layers_per_block, int):
```
Comment (DN6) [inline src/diffusers/models/unets/unet_motion_model.py:373]: I think we can use a similar approach to:
https://github.com/huggingface/diffusers/blob/2f6f426f669ec0ba715ddb941121198ec6ab11d0/src/diffusers/models/unets/unet_2d_condition.py#L415-L419

when setting the reversed blocks if the argument is provided

File: src/diffusers/models/unets/unet_motion_model.py [hunk 6]
```diff
@@ -147,15 +191,20 @@ def __init__(
...
+                transformer_layers_per_block=motion_transformer_per_mid_layers,
             )
         else:
             self.mid_block = None
 
         reversed_block_out_channels = list(reversed(block_out_channels))
         output_channel = reversed_block_out_channels[0]
+
+        reversed_motion_layers_per_block = list(reversed(motion_layers_per_block))
+        reversed_motion_motion_transformer_per_layers = list(reversed(motion_transformer_per_layers))
```
Comment (DN6) [inline src/diffusers/models/unets/unet_motion_model.py:206]:
```suggestion
        reversed_motion_transformer_layers_per_block = list(reversed(motion_transformer_layers_per_block))
```

File: src/diffusers/models/unets/unet_motion_model.py [hunk 7]
```diff
@@ -390,6 +481,16 @@ def __init__(
             else:
                 add_upsample = False
 
+            if reverse_transformer_layers_per_block is not None:
+                curr_layer_transformer = reverse_transformer_layers_per_block[i]
+            else:
+                curr_layer_transformer = reversed_transformer_layers_per_block[i]
```
[the remark this answers is not in the document]
Comment (Arlaz) [reply src/diffusers/models/unets/unet_motion_model.py]: Nice catch, obviously

**Comment by Arlaz** [comment]
The suggested changes have been made, thanks

**Comment by DN6** [comment]
Could we add a fast test for the model here:
https://github.com/huggingface/diffusers/blob/main/tests/models/unets/test_models_unet_motion.py

I think we we should test creating an asymmetric UNetMotionModel. I'd just like to confirm that the updated parameters aren't breaking anything.

**Comment by Arlaz** [comment]
Of course, a test for the most asymmetrical UnetMotionModel possible is added 😃 
It's passing on my side.

**Review by DN6 — APPROVED** [review]

**Comment by DN6** [comment]
Failing tests are unrelated @Arlaz. Merging this.

**Pull request merged**